Mathlib/RingTheory/AdicCompletion/Functoriality.lean

```diff
@@ -0,0 +1,358 @@
+      Submodule.Quotient.mk (p := (I • ⊤ : Submodule R N)) (f x) :=
+  rfl
+
+open AdicCompletion
+
+/-- A linear map induces a linear map on adic cauchy sequences. -/
+@[simps]
+def adicCauchy (f : M →ₗ[R] N) : AdicCauchySequence I M →ₗ[R] AdicCauchySequence I N where
+  toFun a := ⟨fun n ↦ f (a n), fun {m n} hmn ↦ by
+    have hm : Submodule.map f (I ^ m • ⊤ : Submodule R M) ≤ (I ^ m • ⊤ : Submodule R N) := by
+      rw [Submodule.map_smul'']
+      exact smul_mono_right _ le_top
+    apply SModEq.mono hm
+    apply SModEq.map (a.property hmn) f⟩
+  map_add' a b := by ext n; simp
+  map_smul' r a := by ext n; simp
+
+variable (M) in
+theorem adicCauchy_id : adicCauchy I (LinearMap.id (M := M)) = LinearMap.id :=
+  rfl
+
+@[simp]
+theorem adicCauchy_comp (f : M →ₗ[R] N) (g : N →ₗ[R] P) :
+    g.adicCauchy I ∘ₗ f.adicCauchy I = (g ∘ₗ f).adicCauchy I :=
+  rfl
+
+theorem adicCauchy_comp_apply (f : M →ₗ[R] N) (g : N →ₗ[R] P) (a : AdicCauchySequence I M) :
+    g.adicCauchy I (f.adicCauchy I a) = (g ∘ₗ f).adicCauchy I a :=
+  rfl
+
+theorem AdicCompletion.transitionMap_comp_reduceModIdeal (f : M →ₗ[R] N) {m n : ℕ}
+    (hmn : m ≤ n) : transitionMap I N hmn ∘ₗ reduceModIdeal (I ^ n) f =
+      (f.reduceModIdeal (I ^ m) : _ →ₗ[R] _) ∘ₗ transitionMap I M hmn := by
+  ext x
+  simp
+
+/-- `R`-linear version of `adicCompletion`. -/
+private def adicCompletionAux (f : M →ₗ[R] N) :
+    AdicCompletion I M →ₗ[R] AdicCompletion I N :=
+  AdicCompletion.lift I (fun n ↦ reduceModIdeal (I ^ n) f ∘ₗ AdicCompletion.eval I M n)
+    (fun {m n} hmn ↦ by rw [← comp_assoc, AdicCompletion.transitionMap_comp_reduceModIdeal,
+        comp_assoc, transitionMap_comp_eval])
+
+@[simp]
+theorem adicCompletionAux_val_apply (f : M →ₗ[R] N) {n : ℕ} (x : AdicCompletion I M) :
+    (f.adicCompletionAux I x).val n = f.reduceModIdeal (I ^ n) (x.val n) :=
+  rfl
+
+/-- A linear map induces a map on adic completions. -/
+def adicCompletion (f : M →ₗ[R] N) :
+    AdicCompletion I M →ₗ[AdicCompletion I R] AdicCompletion I N where
+  toFun := f.adicCompletionAux I
+  map_add' := by aesop
+  map_smul' r x := by
+    ext n
+    simp only [adicCompletionAux_val_apply, smul_eval, LinearMapClass.map_smul, RingHom.id_apply]
+
+@[simp]
+theorem adicCompletion_val_apply (f : M →ₗ[R] N) {n : ℕ} (x : AdicCompletion I M) :
+    (f.adicCompletion I x).val n = f.reduceModIdeal (I ^ n) (x.val n) :=
+  rfl
+
+/-- Equality of linear maps out of an adic completion can be checked on Cauchy sequences. -/
+@[ext]
+theorem AdicCompletion.ext {f g : AdicCompletion I M → N}
+    (h : ∀ (a : AdicCauchySequence I M),
+      f (AdicCompletion.mk I M a) = g (AdicCompletion.mk I M a)) :
+    f = g := by
+  ext x
+  apply induction_on I M x (fun a ↦ h a)
+
+theorem AdicCompletion.ext' {f g : AdicCompletion I M →ₗ[R] N}
+    (h : ∀ (a : AdicCauchySequence I M),
+      f (AdicCompletion.mk I M a) = g (AdicCompletion.mk I M a)) :
+    f = g := by
+  ext x
+  apply induction_on I M x (fun a ↦ h a)
```

Comment:
Is `AdicCompletion.mk` a linearMap? If so, maybe `(h : f.comp (AdicCompletion.mk I M a) = _)` instead.  See note [partially-applied ext lemmas]

Reply:
Yes it is, I changed it. For the `map_ext''' version, I left it as is for now, because I did not see a way without asking more conditions on `T`. If this is desirable, let me know.